pages/blog/_posts/vercel-cloudflare-wait-until.mdx

```diff
@@ -0,0 +1,126 @@
+  return new Response(data)
+}
+```
+
+The downside of this is that it can slow down the response time of your serverless functions. You ideally want to [keep the critical path of your request minimal](https://www.epicweb.dev/talks/improve-performance-and-reliability-of-your-api-with-events-and-background-functions) so you return the fastest response possible for your users. This is where `waitUntil` comes in.
+
+`waitUntil` allows you to ensure that an async function (a [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)) completes before the serverless function is shut down. As the example code above doesn't require any data returned from `sendMetrics`, we can use `waitUntil` to ensure that the async function completes in the background without blocking the response from returning to the user.
+
+```js
+// ✅ GOOD: The sendMetrics function will execute in the background,
+// but not block the response from returning to the user
+async function POST(request) {
+  const data = await createPost(request.body)
+  waitUntil(sendMetrics({ metric: 'posts.created', increment: 1 }))
+  return new Response(data)
+}
+```
+
+`waitUntil` is a very useful tool that ensures completion, but it does not ensure _success_. If the promise passed to `waitUntil` fails, the serverless function will still complete and return a response. It's important to understand when to use it and when to not use it.
+
+## When to use `waitUntil`
+
+`waitUntil` is useful when you need to perform some asynchronous work that is not critical to the response of the serverless function. This is useful for things like sending metrics, logging, or other things that should not block the user's response. The user should not pay the penalty for these operations, so they can be done in the background.
+
+## When _not_ to use `waitUntil`
+
+As mentioned above, `waitUntil` does not ensure _success_. If the promise passed to `waitUntil` fails, the serverless function will still complete and return a response. This is important to understand when deciding when to use `waitUntil`.
+
+You shouldn't use `waitUntil` if the asynchronous operation effects critical business logic in your application. This is what makes it great for logging and metrics. If the operation fails, it's not the end of the world. But if the operation is critical to the response of the serverless function, you should `await` it instead.
+
+## Can I run background jobs with `waitUntil`?
+
+Short answer: No.
+
+Long answer: `waitUntil` is not designed for running background jobs. There are no retries or tools for handling failures. For running background jobs, you should use a system that handles queueing, retries, and logs failures.
+
+Inngest is designed to handle background jobs reliably. It automatically handles retries, logs, and failures. If you need to run background jobs, you should use Inngest instead of `waitUntil`.
```

Comment:
```suggestion
Inngest is designed to handle background jobs reliably. It automatically handles retries, logs, and failures. If you need to run background jobs, you should consider using Inngest instead of `waitUntil`.
```